Kubeadm unit tests for kubeadm/app/master package

dmmcquay opened the pull request:
Added unit tests for the kubeadm/app/master package testing functionality of tokens.go, kubeconfig.go, manifests.go, pki.go, discovery.go, addons.go, and apiclient.go.

This PR is part of the ongoing effort to add tests (#35025)

/cc @pires @jbeda

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36263)
<!-- Reviewable:end -->

dmmcquay committed:
kubeadm: unit tests for app/master/tokens.go

dmmcquay committed:
kubeadm: unit tests for app/master/kubeconfig.go

dmmcquay committed:
kubeadm: unit tests for app/master/manifests.go

dmmcquay committed:
kubeadm: unit tests for app/master/pki.go

dmmcquay committed:
kubeadm: unit tests for app/master/discovery.go

dmmcquay committed:
kubeadm: unit tests for app/master/addons.go

dmmcquay committed:
kubeadm: unit tests for app/master/apiclient.go

dmmcquay committed:
kubeadm: adding test owner and bazel update

pires commented:
@k8s-bot ok to test

dmmcquay commented:
@pires mind retesting? seems like a bug not related to the PR.

errordeveloper commented:
Is it correct to assume that these tests ran in CI for this PR already?

dmmcquay commented:
I don't know if they ran as part of CI for this PR. But I don't know for sure how the builds are done, it would seem to me that they must have been run.

dmmcquay commented:
`I1104 15:08:00.149] ok  	k8s.io/kubernetes/cmd/kubeadm/app/master	70.425s`
It looks like it did run (grabbed this from jenkins unit/integration logs where it runs other unit tests)

errordeveloper commented on cmd/kubeadm/app/master/manifests_test.go:
```diff
@@ -0,0 +1,577 @@
+		{
+			expected: api.VolumeMount{
+				Name:      "etcd",
+				MountPath: "/var/etcd",
```
I sounds like we should have a constant for this path, but actually what are we testing here exactly? I don't think there is a need to test this really, this function is just a wrapper around structs that are already well tested throughout the code base...

dmmcquay replied:
Its to make sure that nothing gets changed in that function.  I think its a small and simple tests that will ensure that if something got changed this would catch it. Can't see the harm in it.

errordeveloper replied:
Ok, I'd think this should be caught in review, but Kubernetes is a very large project... However, what stops someone from changing the path in the test? If we really have to rely on this path, why don't we leave a comment on the line whete it's defined, telling people to not change it..?

dmmcquay replied:
Valid point, but from a testing perspective, I am just trying to ensure that the functions behave as they meant to and I think this provides an extra catch for someone if they change something.  As to what if someone changed the path in the test function, it would also be reviewed as tests should be reviewed just like standard code. 

Would removing that test make more sense to you then?

pires commented:
As soon as this LGT to @errordeveloper I'd ask for a squash into single commit to keep history simpler.

errordeveloper commented:
To be very honest, to this looks good as long as the CI is happy, and we don't introduce side effects. I am just trying to understand why we need to test misc stuff like etcd path in this fashion, there are so many things that can break at all levels, and we are testing this kind of little thing... Personally, I'd like see more focus on full-blown integration test suite, I don't see how unit tests help kubeadm, but may be I'm not a the right person to ask when it comes to unit tests.

errordeveloper commented:
/lgtm

dmmcquay commented:
@luxas mind taking a look and removing the release note label?

pires commented:
@k8s-bot gce etcd3 e2e test this

pires commented:
@errordeveloper some unit-tests are better than no tests. Having e2e tests is highly desirable but we need to start somewhere, right? Example, remember a bug where the `apiserver` manifest would have two `--etcd-servers`, one pointing to `localhost` and another one pointing to whatever the user had specified with `kubeadm --external-etcd-endpoints`? A unit-test would pick this up without even needing to run e2e, and most probably even an e2e test would fail to point you that.

pires commented:
@k8s-bot test this

pires commented:
@k8s-bot cvm gce e2e test this

dmmcquay committed:
kubeadm: updated pkg node tests to correct type

updated ""k8s.io/kubernetes/pkg/api" to apiv1 "k8s.io/kubernetes/pkg/api/v1" to fix broken types in tests.

dmmcquay commented:
@errordeveloper I had to push up another change, it required a different package for the tests to work, other than that, it has remained the same. Can I get your /lgtm again?

luxas commented:
I'll reapply when it's removed

dmmcquay commented:
@luxas thanks for doing that!

dmmcquay commented:
@luxas looks like more flakes, what do you think? Mind retesting?

luxas commented:
@k8s-bot gci gce e2e test this

luxas commented:
@k8s-bot gce etcd3 e2e test this

luxas commented:
Let's hope tests are passing now and the SQ merges this :+1:

dmmcquay commented:
@luxas looks like another flake, what do you think?
